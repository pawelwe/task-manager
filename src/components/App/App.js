import React, { Component } from 'react';
import uuid from 'uuid';
import toastr from 'toastr';
import {
  findModuleToUpdateIndex,
  sortBy,
  calculateExpiration,
} from '../../utils/utils';

import AddTaskModule from '../AddTaskModule/AddTaskModule';
import TaskModuleList from '../TaskModuleList/TaskModuleList';
import ErrorBoundary from '../ErrorBoundary/ErrorBoundary';

import classes from './App.css';
import wheelIcon from '../../images/wheel-icon.svg';

class App extends Component {
  state = {
    taskModules: [],
  };

  renderTaskModules = () => {
    const taskModules = this.state.taskModules;
    if (taskModules.length < 1) {
      return (
        <div className="container container_content">
          <h5>No task modules...</h5>
          <br />
        </div>
      );
    }
    return (
      <TaskModuleList
        taskModules={taskModules}
        addNewTask={this.handleAddTask}
        toggleTask={this.handleToggleTask}
        removeTask={this.handleRemoveTask}
        removeTaskModule={this.handleRemoveTaskModule}
        sortByPriority={this.handleSortByPriority}
        sortByExpiration={this.handleSortByExpiration}
      />
    );
  };

  handleAddTask = (moduleId, newTaskName, priority, expirationPeriod) => {
    if (!newTaskName || !priority || !expirationPeriod) {
      toastr.info('Please correct your input!');
      return;
    }
    const updatedModules = [...this.state.taskModules];
    const moduleToUpdateIndex = findModuleToUpdateIndex(
      updatedModules,
      moduleId,
    );
    const newTask = {
      name: newTaskName,
      id: uuid.v4(),
      completed: false,
      priority: parseInt(priority, 10),
      creationDate: Date.now(),
      expirationPeriod: parseInt(expirationPeriod, 10),
    };
    updatedModules[moduleToUpdateIndex].tasks.push(newTask);
    this.setState({
      taskModules: updatedModules,
    });
    return updatedModules;
  };

  handleRemoveTask = (moduleId, taskId) => {
    const updatedModules = [...this.state.taskModules];
    const moduleToUpdateIndex = findModuleToUpdateIndex(
      updatedModules,
      moduleId,
    );
    const updatedModuleTasks = updatedModules[
      moduleToUpdateIndex
    ].tasks.filter(task => {
      return task.id !== taskId;
    });
    updatedModules[moduleToUpdateIndex].tasks = updatedModuleTasks;
    this.setState({
      taskModules: updatedModules,
    });
    return updatedModules;
  };

  handleToggleTask = (moduleId, taskId) => {
    const updatedModules = [...this.state.taskModules];
    const moduleToUpdateIndex = findModuleToUpdateIndex(
      updatedModules,
      moduleId,
    );
    const taskToUpdateIndex = updatedModules[
      moduleToUpdateIndex
    ].tasks.findIndex(task => {
      return task.id === taskId;
    });
    updatedModules[moduleToUpdateIndex].tasks[
      taskToUpdateIndex
    ].completed = !updatedModules[moduleToUpdateIndex].tasks[taskToUpdateIndex]
      .completed;
    this.setState({
      taskModules: updatedModules,
    });
    return updatedModules;
  };

  handleAddTaskModule = moduleName => {
    if (!moduleName) {
      toastr.info('Please correct your input!');
      return;
    }
    const taskModules = this.state.taskModules;
    const newModule = {
      title: moduleName,
      id: uuid.v4(),
      tasks: [],
      sortAsc: true,
    };
    const newTaskModules = [...taskModules, newModule];
    this.setState({
      taskModules: newTaskModules,
    });
    return newTaskModules;
  };

  handleRemoveTaskModule = moduleId => {
    const that = this;
    let allModules = [...that.state.taskModules];
    let updatedModules = allModules.filter(module => {
      return module.id !== moduleId;
    });
    const toastrOptions = {
      closeButton: true,
      timeOut: 5000,
      progressBar: true,
      tapToDismiss: false,
      onclick() {
        toastr.clear();
        that.setState({
          taskModules: updatedModules,
        });
      },
      onCloseClick() {
        toastr.clear();
      },
    };
    toastr['info'](
      `<div><span class="CloseDialog">Yes</span></div>`,
      `Do you really want to delete module ${moduleId}?`,
      toastrOptions,
    );
    return updatedModules;
  };

  handleSortByPriority = moduleId => {
    const updatedModules = [...this.state.taskModules];
    const moduleToUpdateIndex = findModuleToUpdateIndex(
      updatedModules,
      moduleId,
    );
    updatedModules[moduleToUpdateIndex].tasks.sort(
      sortBy('priority', updatedModules[moduleToUpdateIndex].sortAsc),
    );
    updatedModules[moduleToUpdateIndex].sortAsc = !updatedModules[
      moduleToUpdateIndex
    ].sortAsc;
    this.setState({
      taskModules: updatedModules,
    });
    return updatedModules;
  };

  handleSortByExpiration = moduleId => {
    const updatedModules = [...this.state.taskModules];
    const moduleToUpdateIndex = findModuleToUpdateIndex(
      updatedModules,
      moduleId,
    );
    updatedModules[moduleToUpdateIndex].sortAsc = !updatedModules[
      moduleToUpdateIndex
    ].sortAsc;
    updatedModules[moduleToUpdateIndex].tasks.sort((a, b) => {
      if (updatedModules[moduleToUpdateIndex].sortAsc) {
        return (
          calculateExpiration(a.creationDate, a.expirationPeriod) -
          calculateExpiration(b.creationDate, b.expirationPeriod)
        );
      } else {
        return (
          calculateExpiration(b.creationDate, b.expirationPeriod) -
          calculateExpiration(a.creationDate, a.expirationPeriod)
        );
      }
    });
    this.setState({
      taskModules: updatedModules,
    });
    return updatedModules;
  };

  render() {
    return (
      <div className={`${classes.App} container`}>
        <ErrorBoundary>
          <header className={classes.App_header}>
            <h1>
              <img
                className={classes.App_logo}
                src={wheelIcon}
                alt="wheel-icon"
              />Task Manager
            </h1>
          </header>
          {this.renderTaskModules()}
          <AddTaskModule addTaskModule={this.handleAddTaskModule} />
        </ErrorBoundary>
      </div>
    );
  }
}

export default App;
