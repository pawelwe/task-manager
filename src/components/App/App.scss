@import 'scss/variables';
@import 'scss/reset';
@import 'scss/_global';
@import 'scss/animations';
@import 'scss/layout';
@import 'scss/toastr';

.App {
  &_header {
    text-align: center;
    margin-bottom: 5rem;
  }

  &_logo {
    display: inline-block;
    vertical-align: middle;
    max-width: 4rem;
    margin-right: 0.6rem;
    margin-top: -5px;
  }
}

@media (min-width: $tablet) {
  .App {
    padding: 2rem;
  }
}

@media (min-width: $desktop) {
  .App {
    padding: 5rem;
  }
}
