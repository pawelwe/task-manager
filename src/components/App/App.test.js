import React from 'react';
import { shallow } from 'enzyme';
import App from './App';

describe('App Component', () => {
  const wrapper = shallow(<App />);
  const instance = wrapper.instance();
  const moduleName = 'Module 1';
  const taskName = 'Task 1';
  let moduleId = null;
  let taskId = null;

  it('Snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('handleAddTaskModule method', () => {
    expect(instance.handleAddTaskModule(moduleName)[0].title).toEqual(moduleName);
  })

  it('handleAddTask method', () => {
    moduleId = instance.renderTaskModules().props.taskModules[0].id;
    expect(instance.handleAddTask(moduleId, taskName, 10, 10)[0].tasks[0].name).toEqual(taskName);
  })

  it('handleToggleTask method', () => {
    taskId = instance.renderTaskModules().props.taskModules[0].tasks[0].id;
    expect(instance.handleToggleTask(moduleId, taskId)[0].tasks[0].completed).toEqual(true);
  })

  it('renderTaskModules method', () => {
    expect(instance.renderTaskModules().props.taskModules[0].title).toEqual(moduleName);
  })

  it('handleSortByProp method', () => {
    instance.handleAddTask(moduleId, 'Task 2', 1, 1);
    expect(instance.handleSortByProp(moduleId, 'priority')[0].tasks[0].priority).toEqual(10);
  })

  it('handleSortByExpiration method', () => {
    expect(instance.handleSortByExpiration(moduleId)[0].tasks[0].expirationPeriod).toEqual(1);
  })

  it('handleRemoveTask method', () => {
    expect(instance.handleRemoveTask(moduleId, taskId)[0].tasks.length).toEqual(1);
  })

  it('handleRemoveTaskModule method', () => {
    expect(instance.handleRemoveTaskModule(moduleId).length).toEqual(0);
  })

});