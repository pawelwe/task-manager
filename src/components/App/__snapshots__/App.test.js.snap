// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component Snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <ErrorBoundary>
        <header
                className={undefined}
        >
                <h1>
                        <img
                                alt="wheel-icon"
                                className={undefined}
                                src="wheel-icon.svg"
                        />
                        Task Manager
                </h1>
        </header>
        <div
                className="container container_content"
        >
                <h5>
                        No task modules...
                </h5>
                <br />
        </div>
        <AddTaskModule
                addTaskModule={[Function]}
        />
</ErrorBoundary>,
      "className": "undefined container",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <header
            className={undefined}
>
            <h1>
                        <img
                                    alt="wheel-icon"
                                    className={undefined}
                                    src="wheel-icon.svg"
                        />
                        Task Manager
            </h1>
</header>,
          <div
            className="container container_content"
>
            <h5>
                        No task modules...
            </h5>
            <br />
</div>,
          <AddTaskModule
            addTaskModule={[Function]}
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h1>
              <img
                            alt="wheel-icon"
                            className={undefined}
                            src="wheel-icon.svg"
              />
              Task Manager
</h1>,
            "className": undefined,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <img
                  alt="wheel-icon"
                  className={undefined}
                  src="wheel-icon.svg"
/>,
                "Task Manager",
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "alt": "wheel-icon",
                  "className": undefined,
                  "src": "wheel-icon.svg",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              "Task Manager",
            ],
            "type": "h1",
          },
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h5>
                No task modules...
</h5>,
              <br />,
            ],
            "className": "container container_content",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "No task modules...",
              },
              "ref": null,
              "rendered": "No task modules...",
              "type": "h5",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "addTaskModule": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <ErrorBoundary>
          <header
                    className={undefined}
          >
                    <h1>
                              <img
                                        alt="wheel-icon"
                                        className={undefined}
                                        src="wheel-icon.svg"
                              />
                              Task Manager
                    </h1>
          </header>
          <div
                    className="container container_content"
          >
                    <h5>
                              No task modules...
                    </h5>
                    <br />
          </div>
          <AddTaskModule
                    addTaskModule={[Function]}
          />
</ErrorBoundary>,
        "className": "undefined container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <header
              className={undefined}
>
              <h1>
                            <img
                                          alt="wheel-icon"
                                          className={undefined}
                                          src="wheel-icon.svg"
                            />
                            Task Manager
              </h1>
</header>,
            <div
              className="container container_content"
>
              <h5>
                            No task modules...
              </h5>
              <br />
</div>,
            <AddTaskModule
              addTaskModule={[Function]}
/>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <h1>
                <img
                                alt="wheel-icon"
                                className={undefined}
                                src="wheel-icon.svg"
                />
                Task Manager
</h1>,
              "className": undefined,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <img
                    alt="wheel-icon"
                    className={undefined}
                    src="wheel-icon.svg"
/>,
                  "Task Manager",
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "alt": "wheel-icon",
                    "className": undefined,
                    "src": "wheel-icon.svg",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "img",
                },
                "Task Manager",
              ],
              "type": "h1",
            },
            "type": "header",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <h5>
                  No task modules...
</h5>,
                <br />,
              ],
              "className": "container container_content",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "No task modules...",
                },
                "ref": null,
                "rendered": "No task modules...",
                "type": "h5",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": "br",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "addTaskModule": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
